include/boost/coroutine2/detail/pull_coroutine.hpp

```diff
@@ -296,17 +296,8 @@ end( pull_coroutine< T > &) {
 
 namespace std {
 
-template< typename T >
-typename boost::coroutines2::detail::pull_coroutine< T >::iterator
-begin( boost::coroutines2::detail::pull_coroutine< T > & c) {
-    return boost::coroutines2::detail::begin( c);
-}
-
-template< typename T >
-typename boost::coroutines2::detail::pull_coroutine< T >::iterator
-end( boost::coroutines2::detail::pull_coroutine< T > & c) {
-    return boost::coroutines2::detail::end( c);
-}
+using boost::coroutines2::detail::begin;
+using boost::coroutines2::detail::end;
```

Comment:
This is still undefined behaviour. You cannot add declarations to namespace `std`.

What are the (undefined) overloads of `std::begin` and `std::end` trying to achieve? To ensure that qualified calls to `std::begin` and `std::end` will use the customized boost ones? Don't do that.

If you want `std::begin` and `std::end` to be valid for a `pull_coroutine` (because it is meant to be container-like) then add `pull_coroutine::begin()` and `pull_coroutine::end()` member functions. If it's not meant to be container-like, then `std::begin(coro)` and `std::end(coro)` should not work with it (since they are constrained to call `coro.begin()` and `coro.end()` respectively).

Reply:
Alternatively, just remove these undefined declarations completely. Require users to use unqualified `begin(coro)` and `end(coro)` calls, which will find the overloads in `boost::coroutines2::detail` using ADL.

Reply:
> I mostly agree but see #31

So fix the docs :shrug: 

That's why I asked about the intended API above. If the point is just to enable range-based `for` then the undefined overloads are not needed. If being able to call `std::begin(coro)` is essential for some reason (even if that's just because it has been documented as supported) then add `coro.begin()` and `coro.end()` member functions, so that the generic `std::begin` and `std::end` can use them. The docs would be lying (they aren't "overloaded") but calling `std::begin(coro)` would still  work, and do the same thing as before. Just without undefined behaviour.

>  I think specializing std::begin/std::end for user-defined types is accepted though.

Yes, but those function templates are not specializations, they are overloads, which are not allowed. And specializing is a bad idea even if it's allowed. http://www.gotw.ca/publications/mill17.htm

Adding names to `std` via a using-declaration is not allowed though, so the change in this PR is still undefined behaviour.

> Also the patch above is backwards compatible with the already released sources (it might not need to be).

But still undefined behaviour.

Reply:
Thank you for the extensive clarification.

Please note the modified pull request which defines .begin and .end members, eliminating the need for new definitions in the std namespace.